sw/host/opentitanlib/src/util/openocd.rs

```diff
@@ -0,0 +1,110 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{Context, Result};
+
+use std::fs::File;
+use std::io::Write;
+use std::process::{Command, Stdio, Child};
+use crate::io::jtag::{Jtag, JtagParams};
+
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+use std::cell::Cell;
+
+use std::net::TcpStream;
+
+use crate::impl_serializable_error;
+
+use std::{thread, time};
+
+
+/// Represents an OpenOCD server and we can interact with interactively
+pub struct OpenOcdClient {
+    /// OpenOCD child process
+    child: Child,
+    /// TcpStream to the telnet port of OpenOCD
+    telnet_stream: Cell<Option<TcpStream>>,
+}
+
+/// Errors related to the OpenOCD client.
+#[derive(Error, Debug, Deserialize, Serialize)]
+pub enum OpenOcdClientError {
+    #[error("No OpenOCD binary path provided")]
+    OpenOcdBinaryRequired,
+    #[error("Generic error {0}")]
+    Generic(String),
+}
+impl_serializable_error!(OpenOcdClientError);
+
+impl OpenOcdClient {
+    pub fn new(opts: &JtagParams) -> Result<OpenOcdClient> {
+        let mut args = Vec::new();
+        if let Some(cfg) = &opts.openocd_adapter_config {
+            args.push("-f");
+            args.push(&cfg);
+        }
+        // Tell OpenOCD how to connect to the lifecycle controller's RISC-V TAP.
+        // This works because lc_ctrl uses an instance of the RISC-V debug
+        // module [0]. Its IRLEN is hardcoded to 5 [1] and its IDCODE is 1 [2].
+        //
+        // [0]: https://docs.opentitan.org/hw/ip/lc_ctrl/doc/#life-cycle-tap-controller
+        // [1]: See where `dmi_jtag_tap` is created in hw/vendor/pulp_riscv_dbg/src/dmi_jtag.sv
+        // [2]: See where `dmi_jtag` is instantiated in hw/ip/lc_ctrl/rtl/lc_ctrl.sv
+        //
+        // Because these commands are command-line args, they *must* end with
+        // semicolons to satisfy the TCL lexer. When a TCL program is read from
+        // a file, newlines are sufficient to separate statements.
+        args.extend(vec![
+            "-c",
+            "transport select jtag;",
+            "-c",
+            "jtag newtap lc_ctrl tap -irlen 5 -expected-id 0x00000001 -ignore-bypass;",
+            "-c",
+            "target create lc_ctrl.tap.0 riscv -chain-position lc_ctrl.tap -rtos hwthread;",
+            "-c",
+            "init;",
+            "-c",
+            "scan_chain;"
+        ]);
+        log::info!("CWD: {:?}", std::env::current_dir());
+        log::info!(
+            "Spawning OpenOCD: {:?} {:?}",
+            opts.openocd,
+            args.join(" ")
+        );
+        let child = Command::new(&opts.openocd)
+            .args(args)
+            .stdin(Stdio::null())
+            .stdout(Stdio::inherit())
+            .stderr(Stdio::inherit())
+            .spawn().context("failed to spawn openocd")?;
```

Comment:
Do we want to spawn the server on `OpenOcdClient` instantiation? Or do we want to have the API include `server_start/stop()` functions for controlling the server state?

I was thinking a bit about your comment [here](https://github.com/lowRISC/opentitan/pull/17590#discussion_r1141477141), specifically the bit about the TAP strap selection. As noted above, it seems our OT chip muxes which TAP is routed to the JTAG pins, rather than daisy chaining the TAPs internally? Therefore, do we want to spawn the openocd server with a different target configuration file (assuming we split the above TAP configs into a separate target config files) depending on the TAP strap configuration? We could do this automatically when you call `server_start()`, and just print which TAP is connected. Wdyt?

Reply:
Yeah my next iteration will probably move the spawning code to `connect()` (which we might want to rename to something else) so that just creating `OpenOcdClient` does nothing. Then `disconnect()` will stop the server so that we can do something like:
```
jtag.disconnect()
change_tap_straps()
maybe_reset_based_on_lifecycle()
jtag.connect()
```

Reply:
awesome, I think that sounds like a good plan.

Reply:
How would you feel about an API like this to avoid the invalid disconnected state?

```rust
struct OpenOCD {
    fn connect() -> Self { /* ... */ }
    fn disconnect(self) { /* ... */ }
}
```

Reply:
Yeah, ignore, it's not worth it